src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.DebugVerifier.cs

```diff
@@ -0,0 +1,211 @@
+            {
+                this.Visit(node.IterationVariableType);
+                this.Visit(node.Expression);
+                // PROTOTYPE(nullable-api): handle the deconstruction
+                //this.Visit(node.DeconstructionOpt);
+                this.Visit(node.Body);
+                return null;
+            }
+
+            public override BoundNode VisitGotoStatement(BoundGotoStatement node)
+            {
+                // There's no need to verify the label children. They do not have types or nullabilities
+                return null;
+            }
+
+            public override BoundNode VisitTypeOrValueExpression(BoundTypeOrValueExpression node)
+            {
+                Visit(node.Data.ValueExpression);
+                return base.VisitTypeOrValueExpression(node);
+            }
+
+            public override BoundNode VisitDynamicCollectionElementInitializer(BoundDynamicCollectionElementInitializer node)
+            {
+                // https://github.com/dotnet/roslyn/issues/33441 dynamic collection initializers aren't being handled correctly
+                VerifyNode(node, overrideSkippedExpression: true);
+                return null;
+            }
+
+            public override BoundNode VisitBinaryOperator(BoundBinaryOperator node)
+            {
+                VisitBinaryOperatorChildren(node);
+                return null;
+            }
+
+            public override BoundNode VisitUserDefinedConditionalLogicalOperator(BoundUserDefinedConditionalLogicalOperator node)
+            {
+                VisitBinaryOperatorChildren(node);
+                return null;
+            }
+
+            private void VisitBinaryOperatorChildren(BoundBinaryOperatorBase node)
+            {
+                // There can be deep recursion on the left side, so verify iteratively to avoid blowing the stack
+                // PROTOTYPE(nullable-api): Handle in rewriter as well
+                while (true)
+                {
+                    VerifyNode(node);
+
+                    Visit(node.Right);
+
+                    if (!(node.Left is BoundBinaryOperatorBase child))
+                    {
+                        Visit(node.Left);
+                        return;
+                    }
+
+                    node = child;
+                }
+            }
+
+            private void VerifyNode(BoundNode node, bool overrideSkippedExpression = false)
+            {
+                if (node is BoundExpression expression && (overrideSkippedExpression || !s_skippedExpression.Contains(expression.Kind)))
+                {
+                    Debug.Assert(_analyzedNullabilityMap.ContainsKey(expression), $"Did not find {expression} `{expression.Syntax}` in the map.");
+                    TypeSymbol.Equals(expression.Type, _analyzedNullabilityMap[expression].Type, TypeCompareKind.ConsiderEverything | TypeCompareKind.IgnoreNullableModifiersForReferenceTypes);
```

Comment:
>TypeSymbol.Equals [](start = 20, length = 17)

`Debug.Assert(TypeSymbol.Equals(...));`?

Reply:
I can actually remove this, it's part of the asserts in `NullableWalker` now (or will be, once it's re-enabled.)

---
In reply to: [265679196](https://github.com/dotnet/roslyn/pull/34017#discussion_r265679196) [](ancestors = 265679196)